posts/progressive-enhancement/index.qmd

```diff
@@ -0,0 +1,152 @@
+    data = x,
+    admission_column = "date_admission",
+    discharge_column = "date_discharge"
+  )
+
+}
+```
+
+If the function was already a generic, then a new method for the new class should be added, leaving everything else unchanged.
+
+## Adding class support to function outputs via progressive enhancement
+
+Adding class support to function outputs is often more challenging.
+A common option is to add a new argument to the function, which would be a boolean indicating whether the output should be of the new class or not.
+But this doesn't fit in the view of progressive enhancement, as it would require users to change their code to benefit from the new classes, or to suffer from breaking changes.
+
+While the new argument approach is sometimes indeed the only possible method, there are some situations where we can have an approach truly following the progressive enhancement philosophy.
+
+In particular, this is the case when the old output was already inheriting from the parent of the new class (hence the importance of carefully choosing the parent class). In this situation, the new attributes from the new class should not interfere with existing code for downstream analysis.
+
+In this case, let's consider a function that was previously returning an unclassed `data.frame` with patient-level data:
+
+```{r}
+create_patient_dataset <- function(n_patients = 10) {
+
+  data <- data.frame(
+    patient_id = seq_len(n_patients),
+    age = sample(18:99, n_patients, replace = TRUE)
+  )
+
+  return(data)
+
+}
+```
+
+We want to start returning a `linelist` object. Because `linelist` objects are `data.frame`s (or `tibble`s) with an extra `attr`, it can be done in a transparent way:
+
+```{r}
+create_patient_dataset <- function(n_patients = 10) {
+
+  data <- data.frame(
+    patient_id = seq_len(n_patients),
+    age = sample(18:99, n_patients, replace = TRUE)
+  )
+
+  data <- linelist::make_linelist(
+    data,
+    id = "patient_id",
+    age = "age"
+  )
+
+  return(data)
+
+}
+
+inherits(data, "data.frame")
+```
+
+For a more realistic example, you can also see the work in progress to integrate the [new `contactmatrix` standard format](https://github.com/socialcontactdata/contactmatrix) for social contact data to the [contactdata package](https://github.com/Bisaloo/contactdata).
+
+This is however only true if code in downstream analysis follows good practices [^1]. If existing code was testing equality of the class to a certain value, it will break when the new class value is appended. This is described in a [post on the R developer blog, when base R was adding a new `array` class value to `matrix` objects](https://developer.r-project.org/Blog/public/2019/11/09/when-you-think-class.-think-again/index.html). Class inheritance should never be tested via `class(x) == "some_class"`. Instead, `inherits(x, "some_class")` or `is(x, "some_class")` should be used to future-proof the code and allow appending an additional in the future.
```

Comment:
```suggestion
This is however only true if code in downstream analysis follows good practices in checking for the class of an object [^1]. If existing code was testing equality of the class to a certain value, it will break when the new class value is appended. This is described in a [post on the R developer blog, when base R was adding a new `array` class value to `matrix` objects](https://developer.r-project.org/Blog/public/2019/11/09/when-you-think-class.-think-again/index.html). Class inheritance should never be tested via `class(x) == "some_class"`. Instead, `inherits(x, "some_class")` or `is(x, "some_class")` should be used to future-proof the code and allow appending an additional in the future.
```